### DISCUSSION TITLE
neopixel led control

### DISCUSSION THREAD

**Discussion by shtaXveed** [digitalninja-ro/klipper-neopixel#5] [General]
Thank you, I've been trying to figure this out for near a week. lol i switched a SR from marlin to klipper after installing an led ring.

**Comment by digitalninja-ro** [comment]
I don't think I understood your message. Can I help you with something?

**Comment by Bostongeorge76** [comment]
Hello, in thanking you for your work, I would like to take this opportunity to ask you a question, I have a led strip which in idle is white, I use the macro NEOPIXEL_DISPLAY LED = "led_name" TYPE = template_type MODE = template_mode for heating and the progress bar but at the end of the print I can't get the color back to white, is there a command to end the macro?

**Reply by burntcoinpurse** [reply]
Same here. I can get them to turn off with a new template and gcode, but when I start a new effect they flicker.
 [display_template led_template_off]
text:
    0,0,0
[gcode_macro NEO_OFF]
gcode:
SET_LED_TEMPLATE LED=my_led TEMPLATE=led_template_off

**Reply by burntcoinpurse** [reply]
I just read through his set up, it's
NEOPIXEL_DISPLAY LED="my_led" TYPE=clear

**Reply by Bostongeorge76** [reply]
> Same here. I can get them to turn off with a new template and gcode, but when I start a new effect they flicker. [display_template led_template_off] text: 0,0,0 [gcode_macro NEO_OFF] gcode: SET_LED_TEMPLATE LED=my_led TEMPLATE=led_template_off

Oh man, thank you so much, it works perfectly now

**Reply by burntcoinpurse** [reply]
Did you make your own or use his beta version with the "clear" command?
I'm using the beta, I also got 2 types to run on 1 LED strip. Top 1/2 for the hotend and the bottom 1/2 for the bed during heating.

**Reply by photodave** [reply]
First off @digitalninja-ro  THANK YOU SO MUCH! for sharing this awesome plugin! Ive only just started using it but Im instantly loving it! :) 

Also @digitalninja-ro and @burntcoinpurse  forgive me as Im an old dog trying to learn new tricks lol. but can you give me some hints as to how you separated the LED's in the code? I have been trying to figure that out as I have a 16 led ring in my stealthburner along with two hot end leds and the logo led and would love to separate this plugin to just run on the ring leds which are on mine at index 1 through 16 and my hotend leds are index 17 & 18 and the logo led at index19. 

I was also looking for a possible way to dim the leds as they are crazy bright. In my previous (very basic) macros for the led's I was running most of them at 0.1 for power except for the hotend LED's which I ran at 0.1 for the camera or if I was in the room I would switch it to 1.0.

Anyway, any help would be greatly appreciated :) Cheers!

**Reply by burntcoinpurse** [reply]
I'm 2 months into learning neopixels and I'm helping solve issues lol. Yeah I have 145 in series and I can do effects like have the top 1/2 light up for the hot end and the bottom for the bed. Both transition from blue to red as they reach target temp. Simultaneously while another portion of the strand is doing a chasing rainbow effect. So yes its very doable.

**Reply by burntcoinpurse** [reply]
This is too much to do on my phone when I get home I'll post how I did it along with the accompanying macros. You will need to install moonraker and https://github.com/julianschill/klipper-led_effect. I basically merged this creators work and julianschill. If you go to that repository, I posted some of my macros in the issue comments. So you can get an idea of how I implemented it.

**Reply by burntcoinpurse** [reply]
@photodave These are my white light macros. I can get my normal LED light to reduce and increase brightness, but I haven't got the neos to yet. I added a high and low setting for you. just change it to your LEDs name and the ones you want to assign. You can run different effects together, on 1 strand in series. 

```
[led_effect light]
leds:
    neopixel:my_led (1-133)
autostart:                          false
frame_rate:                         24
layers:
    static     0  0 top         ( 0.33, 0.33, 0.33)

[led_effect light_low]
leds:
    neopixel:my_led (1-133)
autostart:                          false
frame_rate:                         24
layers:
    static     0  0 top         ( 0.1, 0.1, 0.1)

[led_effect light_high]
leds:
    neopixel:my_led (1-133)
autostart:                          false
frame_rate:                         24
layers:
    static     0  0 top         ( 0.9, 0.9, 0.9)

[gcode_macro LIGHTS_FADE_IN]
gcode:
    SET_LED_EFFECT EFFECT=light FADETIME=5.0
    
[delayed_gcode LIGHTS_FADE_IN_DELAY]
gcode:
    SET_LED_EFFECT EFFECT=light
    
[gcode_macro LIGHTS_FADE_OUT]
gcode:
   SET_LED_EFFECT EFFECT=light STOP=1 FADETIME=5.0
   SET_LED_EFFECT EFFECT=light_low STOP=1 FADETIME=5.0
   SET_LED_EFFECT EFFECT=light_high STOP=1 FADETIME=5.0

[gcode_macro LIGHT_LOW]
gcode:
    SET_LED_EFFECT EFFECT=light_low

[gcode_macro LIGHT_HIGH]
gcode:
    SET_LED_EFFECT EFFECT=light_high

[gcode_macro NEO_OFF]
gcode:
   STOP_LED_EFFECTS
   SET_LED LED=my_led RED=0 GREEN=0 BLUE=0
   #BED_LED LED="my_led" TYPE=clear #my hotend heating display_template off
   #HOTEND_LED LED="my_led" TYPE=clear #my bed heating display_template lights off

```
Then a display config

```
### menu lights ###
[menu __main __lights]
type: list
name: Lights

[menu __main __lights __neooff]
type:command
name: Neo Off
gcode:
    SET_LED LED=my_led RED=0 GREEN=0 BLUE=0
    STOP_LED_EFFECTS
   # BED_LED LED="my_led" TYPE=clear
   # HOTEND_LED LED="my_led" TYPE=clear   (Any custom display_templates can be addded for one easy kill macro)

[menu __main __lights __lightsfadein]
type:command
name: Lights On
gcode:
    LIGHTS_FADE_IN

[menu __main __lights __lightsfadeout]
type:command
name: Lights Off
gcode:
    LIGHTS_FADE_OUT

[menu __main __lights __brightlights]
type:command
name: Bright Lights
gcode:
    LIGHT_HIGH
```

**Reply by burntcoinpurse** [reply]
@photodave https://github.com/digitalninja-ro/klipper-neopixel/discussions/5#discussioncomment-3149015 I explained how I got different effects going at the same time, on only selected LEDs, and how to clear them out.

**Reply by burntcoinpurse** [reply]
I also just added a momentary push button to a spare filament senor port to kill the lights easily. I also tried a click button with a gcode on push and gcode on release. Both work just fine.

**Comment by Sonicboom247** [comment]
Hello, and thanks for this amazing macro, Is it possible to have two macros one after the other? 
would like to see bed/hotend heating then once complete, hand off to print progress. like so
        NEOPIXEL_DISPLAY LED=my_neopixel TYPE=bed_temp MODE=glow
        NEOPIXEL_DISPLAY LED=my_neopixel TYPE=extruder_temp MODE=glow
	NEOPIXEL_DISPLAY LED=my_neopixel TYPE=print_percent MODE=glow

**Reply by burntcoinpurse** [reply]
Why not do them at the same time? But yes you can

**Reply by burntcoinpurse** [reply]
```
SAVE_GCODE_STATE NAME="neopixeldisplay"
NEOPIXEL_DISPLAY LED=my_neopixel TYPE=bed_temp MODE=glow
NEOPIXEL_DISPLAY LED=my_neopixel
TYPE=clear
NEOPIXEL_DISPLAY LED=my_neopixel TYPE=extruder_temp MODE=glow
NEOPIXEL_DISPLAY LED=my_neopixel
TYPE=clear
NEOPIXEL_DISPLAY LED=my_neopixel TYPE=print_percent MODE=glow
RESTORE_GCODE_STATE NAME="neopixeldisplay
```

Then add another
```
NEOPIXEL_DISPLAY LED=my_neopixel
TYPE=clear
```
In your END gcode


Poof 🤯
However you're only going to be able to assign 1 temp display like that if you heat both at the same time. Unless you m190 the bed then m109 the extruder. It'll work, but I like my printer heated within 2 minutes.

**Reply by Sonicboom247** [reply]
ahh!, i think what is causing my issue is i don't have NEOPIXEL_DISPLAY LED=my_neopixel TYPE=clear, but yes i would love to either do them both at the same time the way my Marlin printers handle TEMP_STAT_LED, if i am not mistaken it takes both into consideration, which is probably what this will do. testing soon as print completes.

**Reply by burntcoinpurse** [reply]
If you read the thread before this one it'll take you down that rabbit hole.

**Reply by Sonicboom247** [reply]
Haha! already on it and have my display breathing! LOL. not sure it is working with the top half..lol by top half i mean 
;Lighting
    SAVE_GCODE_STATE NAME="neopixeldisplay"
        NEOPIXEL_DISPLAY LED=Neopixels TYPE=bed_temp MODE=glow
	NEOPIXEL_DISPLAY LED=Neopixels TYPE=clear
	NEOPIXEL_DISPLAY LED=Neopixels TYPE=extruder_temp MODE=glow
	NEOPIXEL_DISPLAY LED=Neopixels TYPE=clear
	NEOPIXEL_DISPLAY LED=Neopixels TYPE=print_percent MODE=glow
	RESTORE_GCODE_STATE NAME="neopixeldisplay"


but my guess is i can replace all that with ?
Heater
leds: Neopixels
Effect Rate:  34   Minimum temperature to activate effect
Cutoff:       0   Disable effect once temp is reached
Palette:         (1.0, 1.0, 0.0) , (0.0, 0.0, 0.0) Color values to blend from Cold to Hot

**Reply by burntcoinpurse** [reply]
https://github.com/julianschill/klipper-led_effect/issues/61#issuecomment-1183284625 
you can edit mine to fit your setup. 
I think I nailed the effect you're going for.

**Reply by burntcoinpurse** [reply]
@Sonicboom247 I would absolutely sugget installing (https://github.com/julianschill/klipper-led_effect) that way you can use the glow effect. This is my beds display_template. If you want to transition you need to set the temp"target or max temp", then the ratio"target temp/actual temp" and then to end if the "ratio" is equal to the "target temp". Put that in a display_template and have a gcode index the template to whatever leds you want. You can try taking out the "mode" and see if it works

```
[display_template led_bed_temp_glow]
text:
{% if printer.heater_bed.target > 0.0 %}
{% set temp = printer.heater_bed.target %}
{% else %}
{% set temp = printer.configfile.settings.heater_bed.max_temp %}
{% endif %}
{% set ratio = printer.heater_bed.temperature / temp %}
{ratio}, 0.0, {1-ratio}, 0.0

[gcode_macro BED_LED]
gcode:
{% set led = params.LED %}
{% set type = params.TYPE %}
{% set mode = params.MODE %}
{% set "my_neopixel" = printer.configfile.settings['neopixel ' ~ led|lower] %}
{% if type == 'clear' %}
    SET_LED_TEMPLATE LED=my_led TEMPLATE=""
    SET_LED LED=my_led RED=0 GREEN=0 BLUE=0 
{% else %}
    {% if mode == 'glow' %}  
        SET_LED_TEMPLATE LED=my_led TEMPLATE=led_bed_temp_glow INDEX=1 TRANSMIT=1
        SET_LED_TEMPLATE LED=my_led TEMPLATE=led_bed_temp_glow INDEX=2 TRANSMIT=1 #"......ect last TRANSMIT needs to =0"
 {% endif %}
{% endif %}

Then I made a macro to run both

[gcode_macro HEAT_LIGHTS_START]
gcode:
BED_LED LED="my_led" TYPE=bed_temp MODE=glow
HOTEND_LED LED="my_led" TYPE=extruder_temp MODE=glow
```

Then a macro to clear
```

[gcode_macro BED_LED_OFF]
gcode:
BED_LED LED="my_led" TYPE=clear
```

**Reply by Sonicboom247** [reply]
Thanks so much for your help, i appreciate it immensely but coming from Marlin and a Windows guy i have to learn how to set temp %, the most i was able to get working after installing julianschill led effects was the breathing by doing the below.. I will not give up, i also need to figure out how to remove the other LED effects i was trying to get working from this original post..lol
[led_effect panel_idle]
autostart: true
frame_rate: 20
leds:
    neopixel:Neopixels
layers:
    breathing  20 1 top (0.5, 0.0, 0.0),(0.6, 0.0, 0.0),(0.7, 0.0, 0.0),(0.8, 0.0, 0.0),(0.9, 0.0, 0.0),(0.6, 0.0, 0.0),(0.7, 0.0, 0.0),(0.8, 0.0, 0.0),(0.9, 0.0, 0.0),(1.0, 0.0, 0.0),(0.0, 0.0, 1.0)

**Reply by burntcoinpurse** [reply]
Read up on DISPLAY_TEMPLATEs and command templates for klipper, and try to make your own. That's what @digitalninja-ro has done here. You don't have to frame it the same. If you look in led_progress.cfg. You'll see he's made a macro to set parameters for display templates. https://www.klipper3d.org/Config_Reference.html?h=display_te#display_template , https://www.klipper3d.org/Command_Templates.html#template-expansion , https://www.klipper3d.org/G-Codes.html#set_led_template

**Reply by photodave** [reply]
Sorry I have been away with work and such. Thank you so very much for your posts, Im going through them now and going to try to implement them :)

**Reply by Sonicboom247** [reply]
please do not apologize, absolutely nothing wrong with the code, this is 100% me. burntcoinpurse has been an immense help.  With his help i was able to get my LED's breathing, showing heating progress and finally print progress.  My only issue now is that breathing is happening while all the rest is happening instead of only at idle, but i am still learning and will figure it out..lol  Mind you, this is on an MKS 12864 V3 display with 3 Neopixels.

**Reply by burntcoinpurse** [reply]
Keep in mind if you use any delayed gcodes or STOP=1 FADETIME=whatevs.  Then you can not stop it. You must wait for it to complete. I'm using a... well it was Ender3V2. Now. It's evolved.  Dual z steppers and a timing belt, bl touch, a 500w silicone bed with silicone springs, pei bed, linear rail on X and dual for Y,  a hermit crab tool change, 3 extruders, a chamber heater, 144 Neopixels, a TFT35 display,  Pi4, 2 adxl345s, a spider hotend, microswiss, a volcano, and a mixing extruder. All ran from an octopus V1.1 (bed using an SSR, obviously) FYI guys, I got it last December and didn't even know how to solder. Now I just finished programing a nano atmega328p to run a neopixel matrix. 3d printers are very expensive gateway drugs lol. 😂

**Reply by Sonicboom247** [reply]
Haha, we are quite alike, eat, sleep, repeat until we get it. My issue is i have 5 printers to maintain, a laser an Openbuilds CNC a Davinci Jr, a resin printer, a Sidewinder X1 with BLtouch on an BTT SKR2, another Sidewinder X1 with an MKS Monster 8 V2, all still on Marlin and the one that brought me to Klipper an artillery Hornet with BLtouch and the Neopixels display. I still have a 5 neopixels to solder up and mount in the frame.

**Reply by burntcoinpurse** [reply]
I wish I had room for all that lmao. I'm tryig to figure out a macro to adjust the brightness of the neopixels in white light from the display. something like  

```
[menu __main __lights __neolight]
type: input
enable: {'neopixel my_led' in printer}
name: Neo Lights: {'%3d' % (menu.input*100)}%
input: {printer['neopixel my_led'].value}
input_min: 0.0
input_max: 1.0
input_step: 0.1
gcode:
    SET_LED LED=my_led RED={menu.input} GREEN={menu.input} BLUE={menu.input}
```

I added a lights menu to my display BTW.
I just wrote this after starting a print so I'll test it tomorrow.
*update* no luck yet, something isn't right, but I know I'm on the right trail.

**Reply by Sonicboom247** [reply]
just a very wild shot in the dark here but wouldn't the input be the encoder?, unless of course it is touch..

**Reply by burntcoinpurse** [reply]
It's a TFT35 E3 V3, but I'm using it in marlin mode I guess (12864 screen mode), Its touch mode does not work on klipper. I have the EXP1, EXP2, and the TFT hooked up to my mother board and addressed in my pinter.cfg. I'm still learning too. So what exactly do you mean? 🤔 

```
#############################################################
# Display
#############################################################

[board_pins]
aliases:
    # EXP1 header
    EXP1_1=PE8, EXP1_2=PE7,
    EXP1_3=PE9, EXP1_4=PE10,
    EXP1_5=PE12, EXP1_6=PE13,    # Slot in the socket on this side
    EXP1_7=PE14, EXP1_8=PE15,
    EXP1_9=<GND>, EXP1_10=<5V>,

    # EXP2 header
    EXP2_1=PA6, EXP2_2=PA5,
    EXP2_3=PB1, EXP2_4=PA4,
    EXP2_5=PB2, EXP2_6=PA7,      # Slot in the socket on this side
    EXP2_7=PC15, EXP2_8=<RST>,
    EXP2_9=<GND>, EXP2_10=PC5

[display]
lcd_type: st7920
cs_pin: PE10
sclk_pin: PE12
sid_pin: PE9
encoder_pins: PB1, PB2
click_pin: ^!PE7
menu_reverse_navigation: true
#kill_pin: 

[output_pin beeper]
pin: PE8
pwm: True
value: 0
shutdown_value: 0
cycle_time: 0.001
scale: 1000
```

**Reply by Sonicboom247** [reply]
disregard, if it is in touch then my theory goes out the door.  I am still learning as well but i have 25 yrs PC experience. which turns out is completely useless at the moment...lol  I was just able to get my TFT which only has 4 pins working on an MKS Monster 8 V2 by connecting it to UART but that is in Marlin still.

**Reply by burntcoinpurse** [reply]
It's not in touch mode. I believe the TFT is all thats actually working, but the beeper and the reset work... which are on UART right? so hmmmmm.... that's where I'm at. lol

**Reply by Sonicboom247** [reply]
not sure about that, my TFT connected directly to the UART 4 pin connector

**Reply by digitalninja-ro** [reply]
> Hello, and thanks for this amazing macro, Is it possible to have two macros one after the other? would like to see bed/hotend heating then once complete, hand off to print progress. like so NEOPIXEL_DISPLAY LED=my_neopixel TYPE=bed_temp MODE=glow NEOPIXEL_DISPLAY LED=my_neopixel TYPE=extruder_temp MODE=glow NEOPIXEL_DISPLAY LED=my_neopixel TYPE=print_percent MODE=glow

Sorry for the late answer, I was in vacation. For what you want I call the progress macro in different parts of my start macro. Take a look down at lines with NEOPIXEL_DISPLAY:

```
[gcode_macro START_PRINT]
gcode:
    {% set BED_TEMP = params.BED_TEMP|default(60)|float %}
    {% set EXTRUDER_TEMP = params.EXTRUDER_TEMP|default(210)|float %}
    CLEAR_PAUSE ; Clears the current paused state, if exist
    M220 S100 ; reset speed
    M104 S160 ; Start preheating
    NEOPIXEL_DISPLAY LED=Pico_Ring TYPE=bed_temp MODE=progress
    M190 S{BED_TEMP} ; Set and wait for bed to reach temperature
    
    G90 ; Use absolute coordinates
    G28 ; Home the printer
    BED_MESH_CALIBRATE ; self explained
    G1 X0 Y0 Z0.2 F5000.0 ; move the nozzle to 0 and wait the extruder
    NEOPIXEL_DISPLAY LED=Pico_Ring TYPE=extruder_temp MODE=progress
    M109 S{EXTRUDER_TEMP} ; set and wait for nozzle to reach temperature
    G92 E0 ; Reset Extruder
    G1 Z2.0 F3000 ; Move Z Axis up little 
    G1 X0.1 Y20 Z0.3 F5000.0 ; Move to start position
    G1 X0.1 Y200.0 Z0.3 F1500.0 E15 ; Draw the first line
    G1 X0.4 Y200.0 Z0.3 F5000.0 ; Move to side a little
    G1 X0.4 Y20 Z0.3 F1500.0 E30 ; Draw the second line
    G92 E0 ; Reset Extruder
    G1 Z2.0 F3000 ; Move Z Axis up little 
    G1 X5 Y20 Z0.3 F5000.0 ; Move over to prevent blob squish
    
    M221 S95 ; set the flow rate
    SKEW_PROFILE LOAD=my_skew_profile ; skew correction
    NEOPIXEL_DISPLAY LED=Pico_Ring TYPE=print_percent MODE=progress
```

**Reply by Sonicboom247** [reply]
Thank you ! that is amazing, and works! between you and Burntcoinpurse, my room is very colorful! I managed to find some RGB strips but they are 12v, I have them in my computer case and work wonderfully, I was wondering if it is possible to add 5 of the LED's to the printer since it has a separate RGB header.  my display is the MKS 12864 V3 and is using pin PB9 for Neopixel and the RGB red is a duplicate on PB9.  is this something you think can be done? if so i can connect a buck converter but prefer not to.

![Screenshot from 2022-07-27 19-57-55](https://user-images.githubusercontent.com/84334417/181392178-ddc702cd-5626-489f-ad99-543bcf3a0e8e.png)

**Reply by digitalninja-ro** [reply]
I'm not a guru in electronics but as far as I know if it's at 12V you should power somehow on 12V.

**Reply by burntcoinpurse** [reply]
RGB leds should be 12v, so check your header on your motherboard with a multimeter to see if its 12v if not just add  a buck converter. Also adding a resistor to an leds negative lead is always a good idea. The value of the resistor depends on what your trying to do vs what you're working with. So you gotta do a little math.

**Comment by bingston** [comment]
Could someone give me a hand.  I'm so glad I came across this.  I'm not a programmer, or coder but I enjoy trying to pick things apart.  I'm also new to Klipper.
I've uploaded the led_progress.cfg to my config folder but I keep getting errors like:
```
Section 'display_template led_extruder_temp_glow' is not a valid config section

Once the underlying issue is corrected, use the "RESTART"
command to reload the config and restart the host software.
Printer is halted
```
I'm trying to do all of this on a Creality Sonic Pad so I'm not sure if that's a factor in this not working.  I've been chasing my tail all day and can't seem to figure it out.  I'd love to get my lights working with Klipper and I'd be grateful for any help.  

Thanks in advance!

**Reply by burntcoinpurse** [reply]
sorry for a late response. Did you get this working? Did you add [include led_progress.cfg] to your printer.cfg ?

**Reply by bingston** [reply]
Thanks for the reply.  It's been a while since I've used that printer, but I think I finally got the lights to respond. The lights just didn't have the same behavior as they did with Marlin.  I'll have to revisit the problem.

**Reply by burntcoinpurse** [reply]
I'm getting notifications on this, so if you need any help, I don't mind.

**Reply by The-Cougar** [reply]
Hello, I’m a rookie and this may be a stupid question. Do you know why when i Import neopixel into my code it still say no module found even though i downloaded and can see it it my downloads?

**Reply by burntcoinpurse** [reply]
Did you add [include led_progress.cfg] to your printer.cfg ?

**Comment by djviper64** [comment]
Bonjour, merci pour cet investissement de temps pour faire des codes pareil. J'ai une question pour une variante, je pilote des leds adressables par wled que j'ai sur mon bureau, mes imprimantes sont au garage, et je voulais savoir si il y avait une possibilité d'adapter la parti progress pour activer des presets par hazard ?

Merci par avance

**Reply by burntcoinpurse** [reply]
This seems doable. Sounds like you'd need a ESP32 synced with your WLED, connected to your printer's motherboard, probably a 3v, It would need to to be put on a spare 3v pin for power, a ground, and a 3vsignal. a spare fan or endstop connection should do. Then you would need to code the ESP32 to play your effect when it receives a signal from the motherboard. So you would need to add this to your firmware. (I.E.) when the print starts set DATAPIN-X to high this would send your signal to your ESP32 triggering the effect or a sequence of effects. This could be easier if you use klipper or a raspberrypi.

**Reply by djviper64** [reply]
J'utilise klipper sur mes deux imprimantes l'une sur raspi 4 et l'autre mks pi ainsi que des ESP8266 j'arrive sans probleme avec des macros intégrées dans le g-code start_print l'allumage des leds quand klipper est connecté à l'imprimante en blanc, la chauffe en rouge, l'impression en cours bleu et la fin d'impression en arc-en-ciel. Mais je voudrais changer l'impression en type bar progresse une led clignote pendant une tranche de 10% une fois arrivée à 11% elle se passe en constante et la suivant commence a clignoter

**Reply by burntcoinpurse** [reply]
https://github.com/julianschill/klipper-led_effect 
Try adding this as well. They've written out a lot of parameters, you should be able to tweak them to get exactly the effect you're going for.

**Comment by The-Cougar** [comment]
PLEASE HELP
I am working on a project that requires adafruit circuitpython neopixel. So I ran

pip3 install neopixel

It said I need to create a virtual environment so I did and ran the command in it and it said successful. I went to my python script to import neopixel

import neopixel Error: No module named ‘neopixel’

So I went back to the terminal to run

pip3 install neopixel

pip install neopixel

pip install adafruit_circuitpython_neopixel

pip install rpi_ws281x

All these commands ran successfully and I ran them in my virtual environment. Then I went back to my script to import neopixel and the same no module found error popped up again.

I then ran pip show rpi_ws281x and then pip show neopixel

Both said they were downloaded. I deleted and reinstalled to try again and the same error occurred after running in my script I looked on the internet for awhile to find solutions and it seemed people as similar issues but no solutions.
 
 Does anybody have a solution?

**Reply by burntcoinpurse** [reply]
This thread is really for using neopixels in klipper, but I love me some Adafruit. What kind of project is it? Does it require a raspberrypi? I can run neopixels using adafruits library with arduino. I usually will use an Attiny85, uno, or an atmega.

**Reply by The-Cougar** [reply]
I am using a raspberry pi to create an audio visualizer with neopixels. Is downloading the adafruit circuit python my best bet in this situation?

**Reply by burntcoinpurse** [reply]
https://a.aliexpress.com/_msfojls 
Here is your easiest option, lol. I use these controllers all the time. These ones only have Bluetooth and sound. But there are similar ones with wifi too. They only need 5v, so you can tie them in easily.
